### ISSUE TITLE
NameError: uninitialized constant possibly due to race in sidekiq lazy server middleware initialization in JRuby

### ISSUE THREAD

**Issue by ivoanjo** [mperham/sidekiq#3043]
Ruby version: jruby-9.1.2.0
Sidekiq / Pro / Enterprise version(s): Sidekiq 4.1.3

I just bootstrapped sidekiq in a skeleton JRuby app. Right now it has a single job

```ruby
    class SidekiqContactsIntake
      include Sidekiq::Worker

      def perform(args)
        puts "Hello from sidekiq: #{args.inspect}"
      end
    end
```

And no custom redis configuration, other than the namespace (and I'm running it with a localhost redis).

I connected a client and submitted some test jobs (with the server offline):
```ruby
100000.times { Sidekiq::Client.push('class' => 'ContactsIntakeWorker::Endpoints::SidekiqContactsIntake', 'args' => ['Hello, world!']) }
```

And then when I run the server sometimes I'm getting the following:

```
2016-06-29T23:43:46.846Z 28759 TID-2io INFO: Booting Sidekiq 4.1.3 with redis options {:namespace=>"contacts-intake-worker", :url=>nil}


         m,
         `$b
    .ss,  $$:         .,d$
    `$$P,d$P'    .,md$P"'
     ,$$$$$bmmd$$$P^'
   .d$$$$$$$$$$P'
   $$^' `"^$$$'       ____  _     _      _    _
   $:     ,$$:       / ___|(_) __| | ___| | _(_) __ _
   `b     :$$        \___ \| |/ _` |/ _ \ |/ / |/ _` |
          $$:         ___) | | (_| |  __/   <| | (_| |
          $$         |____/|_|\__,_|\___|_|\_\_|\__, |
        .d$$                                       |_|

The signal USR1 is in use by the JVM and will not work correctly on this platform
2016-06-29T23:43:46.888Z 28759 TID-2io INFO: Running in jruby 9.1.2.0 (2.3.0) 2016-05-26 7357c8f OpenJDK 64-Bit Server VM 25.91-b14 on 1.8.0_91-8u91-b14-0ubuntu4~15.10.1-b14 +jit [linux-x86_64]
2016-06-29T23:43:46.890Z 28759 TID-2io INFO: See LICENSE and the LGPL-3.0 for licensing details.
2016-06-29T23:43:46.891Z 28759 TID-2io INFO: Upgrade to Sidekiq Pro for more features and support: http://sidekiq.org
2016-06-29T23:43:46.983Z 28759 TID-2io INFO: Starting processing, hit Ctrl-C to stop
2016-06-29T23:43:47.578Z 28759 TID-2js WARN: {"class":"ContactsIntakeWorker::Endpoints::SidekiqContactsIntake","args":["Hello, world!"],"retry":true,"queue":"default","jid":"e53a61c955d2592c6af1d0f4","created_at":1467243773.372,"enqueued_at":1467243773.372}
2016-06-29T23:43:47.579Z 28759 TID-2k0 WARN: {"class":"ContactsIntakeWorker::Endpoints::SidekiqContactsIntake","args":["Hello, world!"],"retry":true,"queue":"default","jid":"c70dca3ba79c7966d403a588","created_at":1467243773.372,"enqueued_at":1467243773.372}
2016-06-29T23:43:47.609Z 28759 TID-2js WARN: NameError: uninitialized constant Sidekiq::Middleware::Server
2016-06-29T23:43:47.612Z 28759 TID-2js WARN: org/jruby/RubyModule.java:3309:in `const_missing'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq.rb:149:in `block in default_server_middleware'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/middleware/chain.rb:81:in `initialize'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq.rb:148:in `default_server_middleware'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq.rb:139:in `server_middleware'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:129:in `block in process'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:168:in `stats'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:128:in `process'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:81:in `process_one'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:68:in `run'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/util.rb:17:in `watchdog'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/util.rb:25:in `block in safe_thread'
2016-06-29T23:43:47.615Z 28759 TID-2iu WARN: {"class":"ContactsIntakeWorker::Endpoints::SidekiqContactsIntake","args":["Hello, world!"],"retry":true,"queue":"default","jid":"15f4159fde2581fdd5371f11","created_at":1467243773.37,"enqueued_at":1467243773.37}
2016-06-29T23:43:47.616Z 28759 TID-2iu WARN: NameError: uninitialized constant Sidekiq::Middleware::Server::RetryJobs
2016-06-29T23:43:47.616Z 28759 TID-2k2 WARN: {"class":"ContactsIntakeWorker::Endpoints::SidekiqContactsIntake","args":["Hello, world!"],"retry":true,"queue":"default","jid":"b7534ce003cb7a7623cadca6","created_at":1467243773.373,"enqueued_at":1467243773.374}
2016-06-29T23:43:47.616Z 28759 TID-2iu WARN: org/jruby/RubyModule.java:3309:in `const_missing'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq.rb:150:in `block in default_server_middleware'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/middleware/chain.rb:81:in `initialize'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq.rb:148:in `default_server_middleware'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq.rb:139:in `server_middleware'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:129:in `block in process'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:168:in `stats'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:128:in `process'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:81:in `process_one'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/processor.rb:68:in `run'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/util.rb:17:in `watchdog'
/home/knuckles/.rvm/gems/jruby-9.1.2.0/gems/sidekiq-4.1.3/lib/sidekiq/util.rb:25:in `block in safe_thread'

# ... repeats many times ...
```
 
This, of course seems to happen only every third time or so.

I grepped the source a bit, and noticed that these classes in particular are being lazy required:
https://github.com/mperham/sidekiq/blob/master/lib/sidekiq.rb#L145-L146
And since there's no synchronization here I'm guessing this is causing a race in loading the classes.

As an experiment, I added a simple call to `Sidekiq.server_middleware` in my initialization code, and in 10 times I tried running sidekiq, I never hit the issue, so it really seems that the issue is that lazy initialization, and that workaround is successful.

If needed, I can attach the whole app, since it's just a bootstrap skeleton.

Thanks for sidekiq! :)

**Comment by mperham** [comment]
You've got the right workaround.  The correct fix is to have server_middleware eagerly called by the CLI code before booting the app.

**Comment by ivoanjo** [comment]
Thanks for the feedback! I guess this warrants adding `Sidekiq.❨╯°□°❩╯︵┻━┻` to my initialization code too.

**Comment by mperham** [comment]
To hit this bug, you need to be running JRuby with a completely vanilla Sidekiq install.  Most people are using Sidekiq Pro/Enterprise or other 3rd party libraries which touch the server middleware during initialization.  No worries though, it'll be fixed for 4.1.4.
